pkg/diagnose/cni.go

```diff
@@ -102,73 +92,83 @@ func detectCalicoConfigMap(clientSet kubernetes.Interface) (bool, error) {
 	return false, nil
 }
 
-func checkCalicoIPPoolsIfCalicoCNI(info *cmd.Cluster) bool {
-	status := cli.NewStatus()
+func checkCalicoIPPoolsIfCalicoCNI(info *cluster.Info, status reporter.Interface) bool {
 
-	found, err := detectCalicoConfigMap(info.KubeClient)
+	status.Start("Trying to detect the Calico ConfigMap")
+	found, err := detectCalicoConfigMap(info.ClientProducer.ForKubernetes())
 	if err != nil {
-		status.Start(fmt.Sprintf("Error trying to detect the Calico ConfigMap: %s", err))
-		status.EndWith(cli.Failure)
+		status.Failure("Error trying to detect the Calico ConfigMap: ", err)
+		status.End()
 
 		return false
 	}
 
 	if !found {
+		status.End()
 		return true
 	}
 
 	status.Start("Calico CNI detected, checking if the Submariner IPPool pre-requisites are configured.")
-
 	gateways, err := info.GetGateways()
 	if err != nil {
-		status.EndWithFailure("Error retrieving Gateways: %v", err)
+		status.Failure("Error retrieving Gateways: ", err)
+		status.End()
+
 		return false
 	}
 
 	if len(gateways) == 0 {
-		status.EndWithWarning("There are no gateways detected on the cluster")
+		status.Warning("There are no gateways detected on the resource")
+		status.End()
+
 		return false
 	}
 
-	client := info.DynClient.Resource(calicoGVR)
+	client := info.ClientProducer.ForDynamic().Resource(calicoGVR)
 
 	ippoolList, err := client.List(context.TODO(), metav1.ListOptions{})
 	if err != nil {
-		status.EndWithFailure("Error obtaining IPPools: %v", err)
+		status.Failure("Error obtaining IPPools: ", err)
+		status.End()
+
 		return false
 	}
 
 	if len(ippoolList.Items) < 1 {
-		status.EndWithFailure("Could not find any IPPools in the cluster")
+		status.Failure("Could not find any IPPools in the resource")
+		status.End()
+
 		return false
 	}
 
 	ippools := make(map[string]unstructured.Unstructured)
 
+	failed := false
 	for _, pool := range ippoolList.Items {
 		cidr, found, err := unstructured.NestedString(pool.Object, "spec", "cidr")
 		if err != nil {
-			status.QueueFailureMessage(fmt.Sprintf("Error extracting field cidr from IPPool %q", pool.GetName()))
+			status.Failure("Error extracting field cidr from IPPool ", pool.GetName())
+			failed = true
 			continue
 		}
 
 		if !found {
-			status.QueueFailureMessage(fmt.Sprintf("No CIDR found in IPPool %q", pool.GetName()))
+			status.Failure("No CIDR found in IPPool ", pool.GetName())
+			failed = true
 			continue
 		}
 
 		ippools[cidr] = pool
 	}
 
-	checkGatewaySubnets(gateways, ippools, status)
+	failed = checkGatewaySubnets(gateways, ippools, status, failed)
```

Comment:
This overwrites the value of `failed`, which may be `true` from the loop above.

Reply:
In more detail, we want `failed` to be `true` if any of the operations failed. In the code above, as we iterate over `ippoolList`, `failed` will be set to `true` if any of the items can’t be used or isn’t found; but the loop continues over the remaining items. When we exit the loop, `failed` will be `false` if everything went OK, `true` if at least one entry wasn’t usable, and `ippools` will only contain the items we can check.

As written above, `failed` will then be overwritten by the result of `checkGatewaySubnets()`, so if the latter succeeds, returning `false`, `failed` will be `false` even if it it was set to `true` earlier. Assuming we do want to call `checkGatewaySubnets` when `failed` is `true` (and I think that’s OK since `ippools` is valid), `failed` should be ored with the result:

```suggestion
	failed = failed || checkGatewaySubnets(gateways, ippools, status, failed)
```

Thus `return !failed` at the end of the function will accurately reflect the result.

Reply:
`checkGatewaySubnets()` returns the value of `failed` as passed. So if `failed` is set to `true` by `ippoolList`, `true` is passed to `checkGatewaySubnets()` which returns `true` if everything is OK. If things in `checkGatewaySubnets()` are not OK, `failed` will be set to `true` (no chance in the state in the current scenario). I think `return failed` at the end should suffice.